src/python/dxpy/nextflow/ImageRef.py

```diff
@@ -0,0 +1,137 @@
+
+    def _cache(self, file_name):
+        """
+        Function to store an image on the platform as a dx file object. Should be implemented in subclasses.
+        :param file_name: A file name under which the image will be saved on the platform
+        :type file_name: String
+        :returns: Tuple[String, String] dx file id, file name (basename)
+        """
+        raise NotImplementedError("Abstract class. Method not implemented. Use the concrete implementations.")
+
+    def _reconstruct_image_ref(self):
+        raise NotImplementedError("Abstract class. Method not implemented. Use the concrete implementations.")
+
+    def _construct_cache_file_name(self):
+        return self._join_if_exists("_", [self._image_name, self._tag])
+
+    @staticmethod
+    def _join_if_exists(delimiter, parts):
+        return delimiter.join([x for x in parts if x])
+
+    def _dx_file_get_name(self):
+        dx_file_handle = DXFile(self._dx_file_id, config["DX_PROJECT_CONTEXT_ID"])
+        return dx_file_handle.describe().get("name")
+
+    def _package_bundle(self):
+        """
+        Function to include a container image stored on the platform into NPA
+        :returns: Dict in the format of {"name": "bundle.tar.gz", "id": {"$dnanexus_link": "file-xxxx"}}
+        """
+        if not self._dx_file_id:
+            cache_file_name = self._construct_cache_file_name()
+            self._dx_file_id = self._cache(cache_file_name)
+        else:
+            cache_file_name = self._dx_file_get_name()
+        return {
+            "name": cache_file_name,
+            "id": {"$dnanexus_link": self._dx_file_id}
+        }
+
+
+class DockerImageRef(ImageRef):
+    def __init__(
+            self,
+            process,
+            digest,
+            dx_file_id=None,
+            repository=None,
+            image_name=None,
+            tag=None
+    ):
+        super().__init__(
+            process,
+            digest,
+            dx_file_id,
+            repository,
+            image_name,
+            tag)
+
+    def _cache(self, file_name):
+        full_image_ref = self._reconstruct_image_ref()
+        docker_pull_cmd = "docker pull {}".format(full_image_ref)
+        docker_save_cmd = "docker save {} | gzip > {}".format(full_image_ref, file_name)
+        for cmd in [docker_pull_cmd, docker_save_cmd]:
+            try:
+                _ = subprocess.check_output(cmd, shell=True)
+            except subprocess.CalledProcessError:
+                err_exit("Failed to run a subprocess command: {}".format(cmd))
+        # may need wait_on_close = True??
+        uploaded_dx_file = upload_local_file(
+            filename=file_name,
+            project=config["DX_PROJECT_CONTEXT_ID"],
+            folder=self._caching_dir,
+            name=file_name,
+            parents=True,
+            properties={"image_digest": self._digest}
+        )
+        return uploaded_dx_file.get_id()
+
+    def _reconstruct_image_ref(self):
```

Comment:
See [spec](https://confluence.internal.dnanexus.com/pages/viewpage.action?pageId=113675686)
TL;DR - it should be referenced by `image_name@digest` (see your own reference, it can't be just digest). In this case tarball name won't have the version, but it will have a digest property. See how `DockerImage` in nextaur solves that.

Reply:
```
        full_ref = self._join_if_exists(":", [repo_and_image_name, self._tag])
        return full_ref
 ```
 
 I am saying here the _tag could be empty, because instead digest is used to render the full_ref together with repo and image name. Like docker.io/library/ubuntu@sha256:26c68657ccce2cb0a31b330cb0be2b5e108d467f641c62e13ab40cbec258c68d


But you only include the case for docker.io/library/ubuntu:22.04

Reply:
yes, but what is the concern here? Is it not covered with tests or this is not explicit in the docs/spec? Could you be more specific in what your expectation is and what you think needs a change?

Reply:
I am worried if the digest format is specified by the user, then we cannot get the full_ref here. 

if the user gives `docker.io/library/ubuntu@sha256:26c68657ccce2cb0a31b330cb0be2b5e108d467f641c62e13ab40cbec258c68d`, then the full_ref generated here is only `docker.io/library/ubuntu`

Reply:
ok, why is it necessary to get a full reference here?

Reply:
fixed [8c73ccd](https://github.com/dnanexus/dx-toolkit/pull/1263/commits/8c73ccd0056d414f4685e6c09b13cefb59639c7d)

Reply:
@YuxinShi0423 please resolve if this is satisfied

Reply:
Thanks! that looks good to me.